src/utils/crt_perf/crt_perf.c

```diff
@@ -0,0 +1,1532 @@
+		   size_t buf_size, struct timespec t)
+{
+	const struct crt_perf_opts *opts = &perf_info->opts;
+	double                      rpc_time;
+	size_t                      loop = (size_t)opts->loop, request_max = opts->request_max,
+	       dir           = (size_t)(opts->bidir ? 2 : 1),
+	       mpi_comm_size = (size_t)perf_info->mpi_info.size;
+
+	rpc_time = d_time2s(t) * 1e6 / (double)(loop * request_max * dir * mpi_comm_size);
+
+	printf("%-*zu%*.*f%*lu\n", 10, buf_size, NWIDTH, NDIGITS, rpc_time, NWIDTH,
+	       (long unsigned int)(1e6 / rpc_time));
+}
+
+void
+crt_perf_print_header_bw(const struct crt_perf_info         *perf_info,
+			 const struct crt_perf_context_info *info, const char *benchmark)
+{
+	const struct crt_perf_opts *opts = &perf_info->opts;
+
+	printf("# CRT %s v" CART_VERSION "\n", benchmark);
+	printf("# %d client process(es)\n", perf_info->mpi_info.size);
+	printf("# Loop %d times from size %zu to %zu byte(s) with %zu request(s) "
+	       "in-flight\n# - %zu bulk transfer(s) per request\n",
+	       opts->loop, opts->buf_size_min, opts->buf_size_max, opts->request_max,
+	       opts->buf_count);
+	if (opts->verify)
+		printf("# WARNING verifying data, output will be slower\n");
+	if (opts->mbps)
+		printf("%-*s%*s%*s\n", 10, "# Size", NWIDTH, "Bandwidth (MB/s)", NWIDTH,
+		       "Time (us)");
+	else
+		printf("%-*s%*s%*s\n", 10, "# Size", NWIDTH, "Bandwidth (MiB/s)", NWIDTH,
+		       "Time (us)");
+	fflush(stdout);
+}
+
+void
+crt_perf_print_bw(const struct crt_perf_info *perf_info, const struct crt_perf_context_info *info,
+		  size_t buf_size, struct timespec t)
+{
+	const struct crt_perf_opts *opts = &perf_info->opts;
+	size_t loop = (size_t)opts->loop, mpi_comm_size = (size_t)perf_info->mpi_info.size,
+	       request_max = opts->request_max, buf_count = opts->buf_count;
+	double avg_time, avg_bw;
+
+	avg_time = d_time2s(t) * 1e6 / (double)(loop * request_max * mpi_comm_size * buf_count);
+	avg_bw = (double)(buf_size * loop * request_max * mpi_comm_size * buf_count) / d_time2s(t);
+
+	if (opts->mbps)
+		avg_bw /= 1e6; /* MB/s, matches OSU benchmarks */
+	else
+		avg_bw /= (1024 * 1024); /* MiB/s */
+
+	printf("%-*zu%*.*f%*.*f\n", 10, buf_size, NWIDTH, NDIGITS, avg_bw, NWIDTH, NDIGITS,
+	       avg_time);
+}
+
+int
+crt_perf_verify_data(const void *buf, size_t buf_size)
+{
+	const char *buf_ptr = (const char *)buf;
+	size_t      i;
+	int         rc;
+
+	for (i = 0; i < buf_size; i++) {
+		CRT_PERF_CHECK_ERROR(buf_ptr[i] != (char)i - 1, error, rc, -DER_INVAL,
+				     "Error detected in bulk transfer, buf[%zu] = %d, "
+				     "was expecting %d!",
+				     i, buf_ptr[i], (char)i);
+	}
+
+	return 0;
+
+error:
+	return rc;
+}
+
+void
+crt_perf_request_complete(const struct crt_cb_info *cb_info)
+{
+	struct crt_perf_request *info = (struct crt_perf_request *)cb_info->cci_arg;
+
+	CRT_PERF_CHECK_ERROR(cb_info->cci_rc != 0, out, info->rc, cb_info->cci_rc,
+			     "callback failed");
+
+	if (info->cb) {
```

Comment:
surprised checkpatch doesnt complain about single statement being wrapped in  { } 's